dev/PushNotifications/PushNotificationsLongRunningTask/NotificationListenerManager.cpp

```diff
@@ -3,42 +3,46 @@
 using namespace Microsoft::WRL;
 using namespace ::ABI::Microsoft::Internal::PushNotifications;
 
-void NotificationListenerManager::Initialize(std::shared_ptr<ForegroundSinkManager> foregroundSinkManager, std::map<std::wstring, std::wstring>& appIdList)
+void NotificationListenerManager::Initialize(std::shared_ptr<ForegroundSinkManager> foregroundSinkManager)
 {
     m_foregroundSinkManager = foregroundSinkManager;
+}
 
+void NotificationListenerManager::SetAppIdMapping(std::map<std::wstring, std::wstring>& appIdList)
+{
     for (auto appData : appIdList)
     {
         AddListener(appData.first, appData.second);
     }
 }
 
-void NotificationListenerManager::AddListener(std::wstring appId, std::wstring processName)
+void NotificationListenerManager::AddListener(std::wstring const& appId, std::wstring const& processName)
 {
-    auto lock = m_lock.lock_exclusive();
-
     THROW_HR_IF(E_INVALIDARG, appId.empty());
     THROW_HR_IF(E_INVALIDARG, processName.empty());
 
-    if (m_notificationListeners.find(appId) == std::end(m_notificationListeners))
-    {
-        ComPtr<INotificationListener> listener;
-        THROW_IF_FAILED(MakeAndInitialize<NotificationListener>(&listener, m_foregroundSinkManager, appId, processName));
-        THROW_IF_FAILED(PushNotifications_RegisterFullTrustApplication(appId.c_str(), GUID_NULL));
-        THROW_IF_FAILED(PushNotifications_RegisterNotificationSinkForFullTrustApplication(appId.c_str(), listener.Get()));
-
-        AgileRef agileListener;
-        THROW_IF_FAILED(AsAgile(listener.Get(), &agileListener));
-        m_notificationListeners.insert({ appId, agileListener });
-    }
+    auto lock = m_lock.lock_exclusive();
+
+    // Make sure we keep the long running sink up-to-date with wpncore.
+    ComPtr<INotificationListener> newListener;
+    THROW_IF_FAILED(MakeAndInitialize<NotificationListener>(&newListener, m_foregroundSinkManager, appId, processName));
+    THROW_IF_FAILED(PushNotifications_RegisterNotificationSinkForFullTrustApplication(appId.c_str(), newListener.Get()));
+
+    // Before updating the listener in the map,
+    // remove any previous listener that belongs to the app.
+    m_notificationListeners.erase(appId);
```

Comment:
Why are we erasing stuff? It isn't necessary. Just replace whatever exists in the map

Reply:
Yes. `insert_or_assign` should be used instead.